linalg/sparsemat.cpp

```diff
@@ -587,16 +616,66 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
    auto d_A = Read(A, nnz);
    auto d_x = x.Read();
    auto d_y = y.ReadWrite();
-   MFEM_FORALL(i, height,
+
+   //Skip if matrix has no non-zeros
+   if (nnz == 0) {return;}
+   if (Device::Allows(Backend::CUDA_MASK) && useCuSparse)
    {
-      double d = 0.0;
-      const int end = d_I[i+1];
-      for (int j = d_I[i]; j < end; j++)
+#ifdef MFEM_USE_CUDA
+      const double alpha = a;
+      const double beta  = 1.0;
+
+      //Initialize once
+      if (!initCuSparse)
       {
-         d += d_A[j] * d_x[d_J[j]];
+         /* create and setup matrix descriptor */
+         cusparseCreateCsr(&matA_descr,Height(), Width(), J.Capacity(),
+                           const_cast<int *>(d_I),
+                           const_cast<int *>(d_J), const_cast<double *>(d_A), CUSPARSE_INDEX_32I,
+                           CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, CUDA_R_64F);
+
+         /*Create handles for input/output vectors */
+         cusparseCreateDnVec(&vecX_descr, x.Size(), const_cast<double *>(d_x),
+                             CUDA_R_64F);
+         cusparseCreateDnVec(&vecY_descr, y.Size(), d_y, CUDA_R_64F);
+
+         /*Create allocate space for kernel */
+         cusparseSpMV_bufferSize(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, &alpha,
```

Comment:
Can you tell me what is the purpose of the `handle`? I have the impression that it's meant to avoid allocating many `dBuffer`, but your code suggests otherwise.

Reply:
This is a pointer type to an opaque cuSPARSE context, which the user must initialize by calling prior to calling cusparseCreate() any other library function.  The dbuffer size should be specific to the sparse matrix no? I would expect matrices of different sizes to require different buffer sizes.

Reply:
But do we need to reallocate memory when the buffer size is smaller? Memory allocations are very expensive, that's why I'm worried. What would be the problem in having a single buffer?

Reply:
For instance, I could see an application that needs to reassemble the matrix at every time step, with the same sparsity, so we could use the same buffer every time.

Reply:
Do you mean share buffers between sparse matrices? Here once ``initCuSparse ;`` is set to true the sparse matrix data structure shouldn't reallocate memory.